src/BookRecommenderSecondary.java

```diff
@@ -0,0 +1,93 @@
+
+        //check if book with sourceTitle exists.
+        if (this.containsBook(sourceTitle)) {
+
+            //map and a queue to store distances, and toVisit
+            Map<String, Integer> distances = new Map1L<>();
+            Queue<String> toVisit = new Queue1L<>();
+
+            //initialize the BFS with the source title and it's attribute
+            distances.add(sourceTitle, 0);
+            toVisit.enqueue(sourceTitle);
+
+            //bfs algorithm i stole from internet then refactored to my purpose
+            while (toVisit.length() > 0) {
+                String currentTitle = toVisit.dequeue();
+
+                if (this.containsBook(currentTitle)) {
+                    Book currentBook = this.getBook(currentTitle);
+                    int currentDist = distances.value(currentTitle);
+
+                    //iterate through all titles to check for adjacency
+                    for (String candidateTitle : bookTitles()) {
+                        //only enqueue if two books are next to each other
+                        if (!distances.hasKey(candidateTitle) && this.containsBook(candidateTitle)) {
+                            Book candidateBook = this.getBook(candidateTitle);
+                            if (this.areAdjacent(currentBook, candidateBook)) {
+                                distances.add(candidateTitle, currentDist + 1);
+                                toVisit.enqueue(candidateTitle);
+                            }
+                        }
+                    }
+                }
+            }
+
+            //use removeAny to find minimal distance
+            int minimalDistance = Integer.MAX_VALUE;
+            Map<String, Integer> temp1 = new Map1L<>();
+            while (distances.size() == 0) {
+                Pair<String, Integer> p = distances.removeAny();
+                String titleKey = p.key();
+                int dist = p.value();
+                // Update minimal distance (ignoring the source).
+                if (!titleKey.equals(sourceTitle) && dist < minimalDistance) {
+                    minimalDistance = dist;
+                }
+                // add the entry to a temporary map to store distance
+                temp1.add(titleKey, dist);
+            }
+            // Restore distances from temp1.
+            distances.transferFrom(temp1);
+
+            //collect books at lowest distance
+            Map<String, Integer> temp2 = new Map1L<>();
+            while (distances.size() == 0) {
+                Pair<String, Integer> p = distances.removeAny();
+                String titleKey = p.key();
+                int dist = p.value();
+                if (!titleKey.equals(sourceTitle) && dist == minimalDistance) {
+                    recommendations.add(titleKey, this.getBook(titleKey));
+                }
+                temp2.add(titleKey, dist);
+            }
+            distances.transferFrom(temp2);
+        }
+
+        return recommendations;
+    }
+}
```

Comment:
Make sure you include `equals()`, `toString()`, and `hashCode()` (or justify the exclusion of any of them). `equals()` is a must for testing.